### PR TITLE
enhancement(gravsearch): Optimise generated SPARQL

### LINKED ISSUE #1464: Gravsearch main query gets values that won't be returned
If a value is used in the WHERE clause of a Gravsearch query, but is not mentioned in the CONSTRUCT clause, the generated main SPARQL query still gets the value. For example:

```sparql
PREFIX knora-api: <http://api.knora.org/ontology/knora-api/simple/v2#>
PREFIX books: <http://0.0.0.0:3333/ontology/00FD/books/simple/v2#>

CONSTRUCT {
    ?book knora-api:isMainResource true .
    ?book books:hasTitle ?title .
    ?book books:hasAuthor ?author .
} WHERE {
    ?book a books:Book .
    ?book books:hasTitle ?title .
    ?book books:hasAuthor ?author .
    ?book books:hasText ?text .
    FILTER knora-api:match(?text, "Paphlagonian")
}
```

Since the text is very large, I would like to optimise the main query by not getting the text from the triplestore.

### LINKED ISSUE #1465: Gravsearch main query gets valueHasStandoff even when not querying standoff
The main query should filter out `knora-base:valueHasStandoff` unless it's actually querying standoff, because if there are a lot of standoff tags, the query result could exceed the maximum result size.

### LINKED ISSUE #1467: Gravsearch query with two matchInStandoff is slow
Example:

```sparql
PREFIX knora-api: <http://api.knora.org/ontology/knora-api/v2#>
PREFIX standoff: <http://api.knora.org/ontology/standoff/v2#>
PREFIX books: <http://0.0.0.0:3333/ontology/00FD/books/v2#>

CONSTRUCT {
    ?book knora-api:isMainResource true .
    ?book books:hasTitle ?title .
    ?book books:hasAuthor ?author .
    ?book books:hasText ?text .
} WHERE {
    ?book a books:Book .
    ?book books:hasTitle ?title .
    ?book books:hasAuthor ?author .
    ?book books:hasText ?text .
    ?text knora-api:valueAsString ?textStr .
    ?text knora-api:textValueHasStandoff ?standoffUnderlineTag .
    ?standoffUnderlineTag a standoff:StandoffUnderlineTag .
    FILTER knora-api:matchInStandoff(?textStr, ?standoffUnderlineTag, "Paphlagonian")
    ?text knora-api:textValueHasStandoff ?standoffItalicTag .
    ?standoffItalicTag a standoff:StandoffItalicTag .
    FILTER knora-api:matchInStandoff(?textStr, ?standoffItalicTag, "soul")
    ?text knora-api:textValueHasStandoff ?standoffOrderedListTag .
    ?standoffOrderedListTag a standoff:StandoffOrderedListTag .
    ?standoffUnderlineTag knora-api:standoffTagHasStartAncestor ?standoffOrderedListTag .
    ?standoffItalicTag knora-api:standoffTagHasStartAncestor ?standoffOrderedListTag .
}
```

### PR DESCRIPTION
**Pull request by benjamingeer**
- [x] In the main query, filter out a value's `valueHasStandoff` in the `UNION` branch that gets the value's other statements. Get the `valueHasStandoff` only in the branch that queries standoff.
- [x] In the prequery, in the implementation of `matchInStandoff`, don't `BIND` a variable for the substring to be searched.

Resolves #1465.
Resolves #1467.

### CODE REVIEW

**Comment by benjamingeer** [comment]
@tobiasschweizer The SPARQL `regex` function is a lot faster if you pass the `substr` function directly as the first argument, instead of making a variable for the substring using `BIND`.

**Comment by benjamingeer** [comment]
@tobiasschweizer This PR just contains a couple of small optimisations that improve the performance of standoff queries in Gravsearch. The more important one is #1464, but I think we can do it in a separate PR.

**Comment by tobiasschweizer** [comment]
This seems an ideal task for my home office day on Thursday. Ok?

**Comment by benjamingeer** [comment]
OK thanks!

File: webapi/src/main/scala/org/knora/webapi/responders/v2/search/gravsearch/mainquery/GravsearchMainQueryGenerator.scala [hunk 1]
```diff
@@ -217,7 +217,14 @@ object GravsearchMainQueryGenerator {
                 StatementPattern.makeInferred(subj = mainAndDependentResourceVar, pred = IriRef(OntologyConstants.KnoraBase.HasValue.toSmartIri), obj = mainAndDependentResourceValueObject),
```
Comment (tobiasschweizer) [inline webapi/src/main/scala/org/knora/webapi/responders/v2/search/gravsearch/mainquery/GravsearchMainQueryGenerator.scala]: > // WHERE patterns for statements about the main and dependent resources' values

Could you change the comment accordingly (no standoff querying here)?

Comment (benjamingeer) [reply webapi/src/main/scala/org/knora/webapi/responders/v2/search/gravsearch/mainquery/GravsearchMainQueryGenerator.scala]: Done in abbabca.

**Comment by benjamingeer** [comment]
Thanks for the review!

**Pull request merged**